blockchain/types/accountkey/account_key_test.go

```diff
@@ -162,3 +168,125 @@ func genAccountKeyRoleBased() AccountKey {
 
 	return NewAccountKeyRoleBasedWithValues(AccountKeyRoleBased{txKey, updateKey, feeKey})
 }
+
+func TestAccountKeyWeightedMultiSig_Validate(t *testing.T) {
+	// declare special block numbers and set hardForkBlockNumberConfig
+	var (
+		blockBeforeHF = uint64(4)
+		blockHF       = big.NewInt(5)
+		blockAfterHF  = uint64(6)
+	)
+	fork.SetHardForkBlockNumberConfig(&params.ChainConfig{IstanbulCompatibleBlock: blockHF})
+
+	// generate multiSigAccount. weights: [1,1,1,1,1,1,1,1,1,1], threshold: 3
+	keys := make(WeightedPublicKeys, 6)
+	for i := 0; i < 6; i++ {
+		k, _ := crypto.GenerateKey()
+		keys[i] = NewWeightedPublicKey(1, (*PublicKeySerializable)(&k.PublicKey))
+	}
+	m := NewAccountKeyWeightedMultiSigWithValues(uint(3), keys)
+
+	// generate test data
+	testData := []struct {
+		recoveredKeys                  []*ecdsa.PublicKey
+		expectValidationResultBeforeHF bool // expectIsValid before istanbul compatible change
+		expectValidationResultAfterHF  bool // expectIsValid values after istanbul compatible change
+	}{
+		// 1. after istanbul compatible change, validation failed because sigNum exceeds keyNum
+		{append(getValidKeys(m, 6), getAnonymousPubKeys(2)...), true, false},
+		{append(getValidKeys(m, 6), getAnonymousPubKeys(4)...), true, false},
+		{append(getValidKeys(m, 6), getValidKeys(m, 2)...), true, false},
+		{append(getValidKeys(m, 6), getValidKeys(m, 4)...), true, false},
+		{append(getValidKeys(m, 2), getValidKeys(m, 6)...), true, false},
+		{append(getValidKeys(m, 1), getValidKeys(m, 6)...), true, false},
+		// 2. after istanbul compatible change, validation failed because invalidSig is included
+		{append(getValidKeys(m, 3), getAnonymousPubKeys(1)...), true, false},
+		{append(getValidKeys(m, 3), append(getValidKeys(m, 2), getAnonymousPubKeys(1)...)...), true, false},
+		{append(getValidKeys(m, 5), getAnonymousPubKeys(1)...), true, false},
+		// 3. validation failed because couldn't reach to the threshold
+		{getValidKeys(m, 1), false, false},
+		{getValidKeys(m, 2), false, false},
+		// 4. result is same between before and after istanbul compatible change
+		{getValidKeys(m, 3), true, true},
+		{getValidKeys(m, 4), true, true},
+		{getValidKeys(m, 6), true, true},
+	}
+
+	// do test
+	for i, tc := range testData {
+		isValid, _ := m.Validate(0, tc.recoveredKeys, common.Address{}, blockBeforeHF)
```

Comment:
Why don't you check returned `validSigNum`?

Reply:
The code is reverted! Thanks for the reviews melvin!!